File: web/hex.js
import * as THREE from "three";

// A flat-top regular hexagon (vertices left/right, flat edges top/bottom) —
// the same orientation Foxhole uses for its hex maps.
export const HEX_RADIUS = 6; // center -> vertex
export const HEX_HEIGHT = 0.6; // extrusion thickness
export const HEX_TOP_Y = HEX_HEIGHT / 2;

// Bounding box of the flat-top hexagon (matches the warapi map-image box).
export const HEX_WIDTH = 2 * HEX_RADIUS; // vertex-to-vertex (x)
export const HEX_DEPTH = Math.sqrt(3) * HEX_RADIUS; // flat-to-flat (z)

// Six flat-top vertices in the XZ plane, angle offset 0 -> vertex at +x and -x.
export function hexCorners(radius = HEX_RADIUS) {
  const pts = [];
  for (let i = 0; i < 6; i++) {
    const a = (Math.PI / 3) * i; // 0,60,...,300
    pts.push(new THREE.Vector2(radius * Math.cos(a), radius * Math.sin(a)));
  }
  return pts;
}

/**
 * Build the hex: an extruded prism body plus a flat top face whose UVs map the
 * warapi bounding box [0..1]^2 onto the hexagon, so a map image lands 1:1 and
 * the cut corners naturally mask the neighbour-bleed at the image corners.
 *
 * Returns { group, top } — `top` is the mesh used as the click/raycast target.
 */
export function createHex({ radius = HEX_RADIUS, height = HEX_HEIGHT, texture = null } = {}) {
  const group = new THREE.Group();
  const corners2d = hexCorners(radius);

  // --- prism body (gives the tile its 3D thickness + side rim) ---
  const shape = new THREE.Shape(corners2d);
  const body = new THREE.ExtrudeGeometry(shape, {
    depth: height,
    bevelEnabled: false,
  });
  // Extrude builds in XY extruded along +Z; rotate so thickness runs along Y.
  body.rotateX(-Math.PI / 2);
  body.translate(0, height, 0); // base at y=0, top at y=height
  body.center(); // recenters all axes; top face ends near y=+height/2

  const bodyMat = new THREE.MeshStandardMaterial({
    color: 0x2b3a2f,
    roughness: 0.9,
    metalness: 0.05,
  });
  const bodyMesh = new THREE.Mesh(body, bodyMat);
  bodyMesh.castShadow = true;
  bodyMesh.receiveShadow = true;
  group.add(bodyMesh);

  // --- top face: a hexagonal ShapeGeometry with bounding-box UVs ---
  const topGeo = new THREE.ShapeGeometry(new THREE.Shape(corners2d));
  applyBoxUVs(topGeo, radius);
  topGeo.rotateX(-Math.PI / 2); // lay flat in XZ (was XY)

  const topMat = texture
    ? new THREE.MeshStandardMaterial({ map: texture, roughness: 0.95, metalness: 0 })
    : new THREE.MeshStandardMaterial({ color: 0x3a5a40, roughness: 0.9 });
  const topMesh = new THREE.Mesh(topGeo, topMat);
  topMesh.position.y = HEX_TOP_Y + 0.001; // sit just above the body's top face
  topMesh.receiveShadow = true;
  group.add(topMesh);

  // --- rim outline for readability ---
  const rim = new THREE.LineSegments(
    new THREE.EdgesGeometry(body),
    new THREE.LineBasicMaterial({ color: 0x10160e })
  );
  group.add(rim);

  group.userData.top = topMesh;
  return { group, top: topMesh };
}

// Map each vertex's (x,y) in the hexagon bounding box to UV [0..1].
// warapi y=0 is the TOP of the map, so flip V to match.
function applyBoxUVs(geometry, radius) {
  const halfW = radius; // x spans [-r, r]
  const halfH = (Math.sqrt(3) / 2) * radius; // y spans [-h, h]
  const pos = geometry.attributes.position;
  const uv = [];
  for (let i = 0; i < pos.count; i++) {
    const x = pos.getX(i);
    const y = pos.getY(i);
    uv.push((x + halfW) / (2 * halfW), (y + halfH) / (2 * halfH));
  }
  geometry.setAttribute("uv", new THREE.Float32BufferAttribute(uv, 2));
}

/**
 * Convert warapi-normalized (nx, ny) in [0,1] (0,0 = top-left of the hex box)
 * to a world position on the hex top face. Shared by markers + image so the
 * data overlay lines up with the texture exactly.
 */
export function normToWorld(nx, ny, lift = 0) {
  const x = (nx - 0.5) * HEX_WIDTH;
  const z = (ny - 0.5) * HEX_DEPTH; // ny=0 -> north (-Z), ny=1 -> south (+Z)
  return new THREE.Vector3(x, HEX_TOP_Y + lift, z);
}

/** Clamp an arbitrary world point onto the hex top (for manual op markers). */
export function onTop(point, lift = 0) {
  return new THREE.Vector3(point.x, HEX_TOP_Y + lift, point.z);
}

File: engine/opengine-core/src/hex.rs
//! Hex geometry constants + coordinate mapping — a 1:1 mirror of `web/hex.js`.
//!
//! `web/hex.js` stays the JS source of truth for runtime placement; this module
//! mirrors the same constants and formulas so Rust-side terrain/wargame math
//! lands on the exact same world points (preserving `normToWorld` alignment).

/// center -> vertex
pub const HEX_RADIUS: f32 = 6.0;
/// extrusion thickness
pub const HEX_HEIGHT: f32 = 0.6;
/// top face Y
pub const HEX_TOP_Y: f32 = HEX_HEIGHT / 2.0;
/// vertex-to-vertex (x): 2 * HEX_RADIUS
pub const HEX_WIDTH: f32 = 2.0 * HEX_RADIUS;
/// flat-to-flat (z): sqrt(3) * HEX_RADIUS. Literal kept in sync with `web/hex.js`
/// (`Math.sqrt(3) * 6`) since `sqrt` is not permitted in a `const`.
pub const HEX_DEPTH: f32 = 10.392_304_845_413_264;

/// Six flat-top hexagon corners in the XZ plane (matches `hexCorners` in hex.js).
/// Angle offset 0 -> a vertex at +x and -x; returned as `[x, y]` pairs.
pub fn hex_corners(radius: f32) -> [[f32; 2]; 6] {
    let mut pts = [[0.0_f32; 2]; 6];
    for (i, p) in pts.iter_mut().enumerate() {
        let a = (std::f32::consts::PI / 3.0) * i as f32; // 0,60,...,300
        *p = [radius * a.cos(), radius * a.sin()];
    }
    pts
}

/// Convert warapi-normalized `(nx, ny)` in `[0,1]` (`(0,0)` = top-left / NW of the
/// hex box) to a world position `[x, y, z]` on the hex top face. Mirror of
/// `normToWorld` in hex.js so overlays line up with the texture exactly.
pub fn norm_to_world(nx: f32, ny: f32, lift: f32) -> [f32; 3] {
    let x = (nx - 0.5) * HEX_WIDTH;
    let z = (ny - 0.5) * HEX_DEPTH; // ny=0 -> north (-Z), ny=1 -> south (+Z)
    [x, HEX_TOP_Y + lift, z]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn depth_matches_sqrt3() {
        assert!((HEX_DEPTH - 3.0_f32.sqrt() * HEX_RADIUS).abs() < 1e-4);
    }

    #[test]
    fn center_maps_to_origin() {
        let [x, _y, z] = norm_to_world(0.5, 0.5, 0.0);
        assert!(x.abs() < 1e-6 && z.abs() < 1e-6);
    }

    #[test]
    fn corners_span_bounding_box() {
        let c = hex_corners(HEX_RADIUS);
        let max_x = c.iter().map(|p| p[0]).fold(f32::MIN, f32::max);
        assert!((max_x - HEX_RADIUS).abs() < 1e-5);
    }
}
